components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java

```diff
@@ -580,6 +592,224 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
+     * request query parameters.
+     * i.e. Adaptive Script example:
+     * This will ignore any other definition (common, local) of the authenticatorParams.
+     * var onLoginRequest = function(context) {
+     * executeStep(1, {
+     * authenticatorParams: {
+     * federated: {
+     * "Google Calender": {
+     * federated_token_scope: "https://googleapis.calander.readonly https://google.calander.list"
+     * }}}}, {});}
+     * i.e Authorize request query param example:
+     * /authorize?response_type=id_token&client_id={ClientId}&redirect_uri={https://app/callback}
+     * &scope=email profile openid
+     * &federated_token_scope=Google Calender;read write,Microsoft Authenticator;https://googleapis.calender
+     *
+     * @param context                 The authentication context.
+     * @param authenticatorProperties The authenticator properties.
+     * @param scopes                  The scopes defined in the authenticator properties.
+     * @return The IDP defined scope and the validated scopes requested by the application.
+     */
+    private String addValidScopesForFederatedTokenSharing(AuthenticationContext context,
+                                                          Map<String, String> authenticatorProperties, String scopes) {
+        /*
+        The scopes for the federated tokens are evaluated only of the authenticator configuration ShareFederatedToken
+        is enabled and the application has requested the federated token.
+         */
+        if (!Boolean.parseBoolean(authenticatorProperties.get(SHARE_FEDERATED_TOKEN_CONFIG)) ||
+                !requestedToShareFederatedToken(context)) {
+            return scopes;
+        }
+
+        // Get the application requested scopes for the federated tokens.
+        String requestedScopesForTokenSharing = getRequestedScopesForTokenSharing(context);
+
+        // Validating the application requested scopes by the authenticator allowed scopes for federated token sharing.
+        String validScopesForTokenSharing = validateScopeForTokenSharing(
+                authenticatorProperties.get(OIDCAuthenticatorConstants.FEDERATED_TOKEN_ALLOWED_SCOPE),
+                requestedScopesForTokenSharing);
+
+        if (StringUtils.isNotBlank(validScopesForTokenSharing)) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Valid scopes matched for federated token sharing: " + validScopesForTokenSharing);
+            }
+            scopes = (scopes + SPACE + validScopesForTokenSharing).trim();
+        }
+        return scopes;
+    }
+
+    /**
+     * This method returns the scopes requested by the application for the federated tokens.
+     *
+     * @param context The authentication context.
+     * @return The scopes requested by the application for token sharing.
+     */
+    private String getRequestedScopesForTokenSharing(AuthenticationContext context) {
```

Comment:
missing new line after the method declaration

Reply:
Fixed 473c2f0d7c5775dee2cb9ced70449de004c3184f